### ISSUE TITLE
One-off backfill script for Walters Museum data

### ISSUE THREAD

**Issue by AetherUnbound** [WordPress/openverse-catalog#772]
## Description

We will need a one-off crawling script which will run against the data that we have stored in the catalog from the provider `waltersartmuseum` to acquire image `height`, `width`, `filesize`, and `filetype` information from each of the still-available images. This information will be used to update the records in the catalog, and is necessary for the data normalization steps we wish to perform (see #536 and #647).

This should be able to be accomplished using a polite crawler which visits all of the `url` values we have for Walters data. I don't believe we'll be able to use `HEAD` requests only since we will actually need the image data itself to determine `height` and `width`:

```python
[ins] In [1]: import requests

[ins] In [2]: r = requests.head("https://static.thewalters.org/images/PS1_37.1513_Fnt_DD_T11.jpg")

[ins] In [4]: r.headers
Out[4]: {'Content-Length': '471138', 'Content-Type': 'image/jpeg', 'Last-Modified': 'Mon, 18 Nov 2013 23:36:26 GMT', 'Accept-Ranges': 'bytes', 'ETag': '"1cc97fffb6e4ce1:0"', 'Server': 'Microsoft-IIS/7.5', 'X-Powered-By': 'ASP.NET', 'Date': 'Wed, 05 Oct 2022 21:26:36 GMT'}
```

## Additional context

This rationale for this decision can be found on our Make WP blog: [Next steps for Walters Art Museum data](https://make.wordpress.org/openverse/2022/09/30/next-steps-for-walters-art-museum-data/).

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

**Comment by obulat** [comment]
Would the data be saved in a TSV format? We would probably need 5 columns: `id`, `height`, `width`, `filesize`, and `filetype`. Would adding `foreign_landing_url` or some other column help increase the speed of writing the data to the upstream table later?
Is writing the data collected using this script also a part of this issue, or do we need a new one specifically for writing a sub-sample of TSV columns to the database? Would we use a query like `UPDATE image WHERE id=id SET height=height ...`?

**Comment by AetherUnbound** [comment]
Hmm, good questions! I suppose since this is a one-off backfill, it might make sense to also have a one-off script which we could run to populate the values in the catalog. That way we could test it as well to make sure the final data is what we expect. We could perform a similar set of steps to the upsert we do for ingestion: Load the TSV into a temporary table, then update all of those columns based on matching IDs. 

If we're matching on an index (like `identifier`), I don't think having multiple indices would speed up the insert. That's just my instinct though, I don't have any data to back that assumption.

**Comment by dylan-k** [comment]
Hello, my name is Dylan Kinnett and I'm the manager of web at the Walters Art Museum. I noticed this issue on github, related to our api and static servers. Since those servers are undergoing some changes, potentially breaking changes, I wanted to reach out to you about this open and ongoing work. Feel free to send me an email so we can chat offline from this thread if needed: dkinnett@thewalters.org

**Comment by AetherUnbound** [comment]
Hi @dylan-k! We're so glad that you reached out 😄 We're happy to discuss here, unless there might be any confidential information that would need to be shared in our exchange. We do plan on keeping the existing Walters Museum data in Openverse indefinitely, but it would be fantastic to be able to continue updating it if that option is available from the Walters Museum API. We tried doing so as you might have seen in this issue (https://github.com/WaltersArtMuseum/api-thewalters-org/issues/75#issuecomment-1264031813). It seems like this might be the best issue for us to track the issues with the API upstream, is that correct? https://github.com/WaltersArtMuseum/api-thewalters-org/issues/76

We're more than happy to test your API once you're able to make those changes to the servers!

**Comment by dylan-k** [comment]
Hello @AetherUnbound 

Yes we would also love to have your data keep up-to-date with ours. We strive to share any additions or updates to the data on a nightly basis. This can include the addition of new artworks whose info. is now complete enough to go online, or additional images of artworks already online, or even new information with discoveries about the artwork.

The trouble is, as you saw, the architecture behind our API is essentially obsolete. We're going to have to take the API offline for a while, to rebuild it. I don't have an exact schedule for the project, and I'm not certain what, if any, changes this would entail for the data (hopefully few/none).

For the meanwhile, would it help if we could share the data with you another way? Then, hopefully after the API is back, it could be used to catch up with any changes?

**Comment by AetherUnbound** [comment]
@dylan-k yes! We've been working on expanding the ways that we can ingest data, what form would you be able to provide the data in? Similarly I don't know that we could give a timeline on altering our pipelines to handle the new format, but it'd certainly be a way to unblock this work 🙂

**Comment by dylan-k** [comment]
.csv would probably be easiest. Would that work for you?

**Comment by AetherUnbound** [comment]
I think so, @WordPress/openverse-catalog I think we might be able to use a similar bulk import approach to iNaturalist to this for the time being. What do you all think?

**Comment by dylan-k** [comment]
@AetherUnbound it's a good idea to have the CSV in a way you can check it for updates. I'll set it up that way, published where you can access it from one of the museum's web servers. As an update, I've got the queries ready to produce the data. I'm having a colleague review the queries and should be able to publish something you can use, later this month.
